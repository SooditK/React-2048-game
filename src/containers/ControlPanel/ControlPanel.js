import React, { Component } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';

import WrapperButton from '../../components/WrapperButton';
import Speaker from '../../components/Speaker';
import styles from './controlPanel.scss';
import resetSvg from '../../assets/svg/reset.svg';
import undoSvg from '../../assets/svg/undo.svg';
import arrowSvg from '../../assets/svg/arrow.svg';

const keyUp = 38;
const keyRight = 39;
const keyDown = 40;
const keyLeft = 37;
const keyW = 87;
const keyS = 83;
const keyA = 65;
const keyD = 68;
const keyN = 78;
// const keySpace = 32;

// Button and keyboard control panel
export default class ControlPanel extends Component {
  static propTypes = {
    delay: PropTypes.number.isRequired,
    onMoveUp: PropTypes.func.isRequired,
    onMoveDown: PropTypes.func.isRequired,
    onMoveLeft: PropTypes.func.isRequired,
    onMoveRight: PropTypes.func.isRequired,
    onPlaceRandom: PropTypes.func.isRequired,
    onReset: PropTypes.func.isRequired,
    onUndo: PropTypes.func.isRequired,
    isMoved: PropTypes.bool.isRequired,
    pastLen: PropTypes.number.isRequired,
    audioMove: PropTypes.instanceOf(Audio).isRequired,
    audioPopup: PropTypes.instanceOf(Audio).isRequired
  };

  constructor(...args) {
    super(...args);

    this.state = {
      speakerOn: true
    };

    const { delay } = this.props;
    this.keyUpHandler = debounce(this.handleKeyUp, delay, {
      leading: true
    });
  }

  componentWillMount() {
    document.addEventListener('keyup', this.keyUpHandler, false);

    // Disable arrow keys scroll page
    document.addEventListener('keydown', this.keyDownHandler, false);
  }

  componentWillUnmount() {
    // Never forget remove event after component unmounted,
    // avoid memory leak
    document.removeEventListener('keyup', this.keyUpHandler, false);
    document.removeEventListener('keydown', this.keyDownHandler, false);
  }

  keyDownHandler = e => {
    switch (e.keyCode) {
      case keyUp:
      case keyDown:
      case keyLeft:
      case keyRight:
        // case keySpace:
        e.preventDefault();
        break;
      default:
        break;
    }
  };

  handleKeyUp = e => {
    switch (e.keyCode) {
      case keyW:
      case keyUp:
        this.handleMoveUp();
        break;
      case keyS:
      case keyDown:
        this.handleMoveDown();
        break;
      case keyA:
      case keyLeft:
        this.handleMoveLeft();
        break;
      case keyD:
      case keyRight:
        this.handleMoveRight();
        break;
      case keyN:
        this.props.onReset();
        break;
      default:
        break;
    }
  };

  handleSpeakerClick = speakerOn => {
    this.setState({
      speakerOn
    });
  };

  generalMove = func => {
    func();
    const { isMoved } = this.props;
    const { speakerOn } = this.state;
    if (speakerOn && isMoved) {
      this.props.audioMove.play();
    }
    // Popup new number after delay a short time
    setTimeout(() => {
      this.props.onPlaceRandom();
      if (speakerOn && isMoved) {
        this.props.audioPopup.play();
      }
    }, 300);
  };

  handleMoveUp = () => {
    this.generalMove(this.props.onMoveUp);
  };

  handleMoveDown = () => {
    this.generalMove(this.props.onMoveDown);
  };

  handleMoveLeft = () => {
    this.generalMove(this.props.onMoveLeft);
  };

  handleMoveRight = () => {
    this.generalMove(this.props.onMoveRight);
  };

  handleUndo = () => {
    if (this.props.pastLen > 3) {
      this.props.onUndo(-2);
    }
  };

  render() {
    const { delay } = this.props;
    return (
      <div className={styles.panel}>
        <div className={styles.rowBtn}>
          <div className={styles.btn}>
            <Speaker onClick={debounce(this.handleSpeakerClick, delay)} />
          </div>
          <WrapperButton onClick={debounce(this.handleUndo, delay)}>
            <img src={undoSvg} alt="undo" />
          </WrapperButton>
          <WrapperButton onClick={debounce(this.props.onReset, delay)}>
            <img src={resetSvg} alt="reset" />
          </WrapperButton>
        </div>
        <WrapperButton
          type="primary"
          onClick={debounce(this.handleMoveUp, delay)}
        >
          <img src={arrowSvg} className={styles.up} alt="arrow up" />
        </WrapperButton>
        <div className={styles.arrows}>
          <WrapperButton
            type="primary"
            onClick={debounce(this.handleMoveLeft, delay)}
          >
            <img src={arrowSvg} className={styles.left} alt="arrow left" />
          </WrapperButton>
          <WrapperButton
            type="primary"
            onClick={debounce(this.handleMoveDown, delay)}
          >
            <img src={arrowSvg} className={styles.down} alt="arrow down" />
          </WrapperButton>
          <WrapperButton
            type="primary"
            onClick={debounce(this.handleMoveRight, delay)}
          >
            <img src={arrowSvg} className={styles.right} alt="arrow right" />
          </WrapperButton>
        </div>
      </div>
    );
  }
}
