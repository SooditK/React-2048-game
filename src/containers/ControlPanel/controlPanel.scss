@import '../../assets/styles/vars';

.panel {
  align-items: center;
  display: flex;
  flex: 1;
  flex-direction: column;

  .row-btn {
    align-items: center;
    display: flex;
    flex-grow: 1;
  }

  .btn {
    margin-right: 0.77vw;
    margin-top: 0.77vw;
  }

  img {
    margin-bottom: -0.35vw;
    width: 2.1vw;
  }

  .arrows {
    display: flex;
    flex-direction: row;
  }

  .left {
    transform: rotate(270deg);
  }

  .right {
    transform: rotate(90deg);
  }

  .down {
    transform: rotate(180deg);
  }
}
