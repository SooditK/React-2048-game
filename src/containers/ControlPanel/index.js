import { connect } from 'react-redux';
import { ActionCreators } from 'redux-undo';
import {
  moveUp,
  moveDown,
  moveLeft,
  moveRight,
  placeRandom,
  reset
} from '../../reducers/board';
import ControlPanel from './ControlPanel';

const mapStateToProps = state => ({
  isMoved: state.present.board.present.isMoved,
  pastLen: state.past.length
});
const mapDispatchToProps = {
  onPlaceRandom: placeRandom,
  onMoveUp: moveUp,
  onMoveDown: moveDown,
  onMoveLeft: moveLeft,
  onMoveRight: moveRight,
  onReset: reset,
  onUndo: ActionCreators.jump // Undo move and generated cell
};

export default connect(mapStateToProps, mapDispatchToProps)(ControlPanel);
