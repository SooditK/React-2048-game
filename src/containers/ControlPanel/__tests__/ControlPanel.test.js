import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import ControlPanel from '../ControlPanel';

Enzyme.configure({ adapter: new Adapter() });
const audio = new Audio();

describe('<ControlPanel />', () => {
  it('component render', () => {
    const fn = () => {};
    const panel = renderer.create(
      <ControlPanel
        delay={1}
        onMoveUp={fn}
        onMoveDown={fn}
        onMoveLeft={fn}
        onMoveRight={fn}
        onPlaceRandom={fn}
        onReset={fn}
        onUndo={fn}
        isMoved={false}
        pastLen={1}
        audioMove={audio}
        audioPopup={audio}
      />
    );
    expect(panel.toJSON()).toMatchSnapshot();
  });

  it('click and keyboard events', () => {
    const moveUp = jest.fn();
    const moveDown = jest.fn();
    const moveLeft = jest.fn();
    const moveRight = jest.fn();
    const placeRandom = jest.fn();
    const reset = jest.fn();
    const undo = jest.fn();
    const div = document.createElement('div');
    const panel = mount(
      <ControlPanel
        delay={0}
        onMoveUp={moveUp}
        onMoveDown={moveDown}
        onMoveLeft={moveLeft}
        onMoveRight={moveRight}
        onPlaceRandom={placeRandom}
        onReset={reset}
        onUndo={undo}
        isMoved
        pastLen={4}
        audioMove={audio}
        audioPopup={audio}
      />,
      { attachTo: div }
    );
    div.dispatchEvent(new Event('focus'));
    // div.dispatchEvent(new KeyboardEvent("keypress", { key: "ArrowUp" }));
    panel.find('[alt="arrow up"]').simulate('click');
    // Fast-forward until all timers have been executed
    jest.runAllTimers();
    expect(moveUp.mock.calls.length).toBe(1);
    expect(placeRandom.mock.calls.length).toBe(1);

    panel.find('[alt="arrow down"]').simulate('click');
    jest.runAllTimers();
    expect(moveDown.mock.calls.length).toBe(1);
    expect(placeRandom.mock.calls.length).toBe(2);

    panel.find('[alt="arrow left"]').simulate('click');
    jest.runAllTimers();
    expect(moveLeft.mock.calls.length).toBe(1);
    expect(placeRandom.mock.calls.length).toBe(3);

    panel.find('[alt="arrow right"]').simulate('click');
    jest.runAllTimers();
    expect(moveRight.mock.calls.length).toBe(1);
    expect(placeRandom.mock.calls.length).toBe(4);

    const ins = panel.instance();

    panel.find('[alt="undo"]').simulate('click');
    jest.runAllTimers();
    expect(undo.mock.calls.length).toBe(1);
    // Expect first call first argument to be -2
    expect(undo.mock.calls[0][0]).toBe(-2);
    panel.setProps({ pastLen: 1 });
    panel.find('[alt="undo"]').simulate('click');
    expect(undo.mock.calls.length).toBe(1);

    ins.handleSpeakerClick(false);
    expect(ins.state.speakerOn).toBe(false);

    ins.handleKeyUp({ keyCode: 1 });
    const preventDefault = jest.fn();
    ins.keyDownHandler({ keyCode: 1 });
    ins.keyDownHandler({ keyCode: 38, preventDefault });
    expect(preventDefault.mock.calls.length).toBe(1);

    panel.find('[alt="reset"]').simulate('click');
    jest.runAllTimers();
    expect(reset.mock.calls.length).toBe(1);

    ins.handleKeyUp({ keyCode: 78 });
    jest.runAllTimers();
    expect(reset.mock.calls.length).toBe(2);

    ins.handleKeyUp({ keyCode: 38 });
    ins.handleKeyUp({ keyCode: 87 });
    jest.runAllTimers();
    expect(moveUp.mock.calls.length).toBe(3);
    expect(placeRandom.mock.calls.length).toBe(6);

    ins.handleKeyUp({ keyCode: 39 });
    ins.handleKeyUp({ keyCode: 68 });
    jest.runAllTimers();
    expect(moveRight.mock.calls.length).toBe(3);
    expect(placeRandom.mock.calls.length).toBe(8);

    ins.handleKeyUp({ keyCode: 40 });
    ins.handleKeyUp({ keyCode: 83 });
    jest.runAllTimers();
    expect(moveDown.mock.calls.length).toBe(3);
    expect(placeRandom.mock.calls.length).toBe(10);

    ins.handleKeyUp({ keyCode: 37 });
    ins.handleKeyUp({ keyCode: 65 });
    jest.runAllTimers();
    expect(moveLeft.mock.calls.length).toBe(3);
    expect(placeRandom.mock.calls.length).toBe(12);
  });
});
