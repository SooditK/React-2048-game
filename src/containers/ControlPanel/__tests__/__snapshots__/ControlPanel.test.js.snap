// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ControlPanel /> component render 1`] = `
<div
  className="panel"
>
  <div
    className="rowBtn"
  >
    <div
      className="btn"
    >
      <button
        className="btn-flat btn-md btn-default"
        onClick={[Function]}
      >
        <img
          alt="speaker"
          src="speaker-on.svg"
        />
      </button>
    </div>
    <div
      className="btn "
    >
      <button
        className="btn-flat btn-md btn-default"
        onClick={[Function]}
      >
        <img
          alt="undo"
          src="undo.svg"
        />
      </button>
    </div>
    <div
      className="btn "
    >
      <button
        className="btn-flat btn-md btn-default"
        onClick={[Function]}
      >
        <img
          alt="reset"
          src="reset.svg"
        />
      </button>
    </div>
  </div>
  <div
    className="btn "
  >
    <button
      className="btn-flat btn-md btn-primary"
      onClick={[Function]}
    >
      <img
        alt="arrow up"
        className="up"
        src="arrow.svg"
      />
    </button>
  </div>
  <div
    className="arrows"
  >
    <div
      className="btn "
    >
      <button
        className="btn-flat btn-md btn-primary"
        onClick={[Function]}
      >
        <img
          alt="arrow left"
          className="left"
          src="arrow.svg"
        />
      </button>
    </div>
    <div
      className="btn "
    >
      <button
        className="btn-flat btn-md btn-primary"
        onClick={[Function]}
      >
        <img
          alt="arrow down"
          className="down"
          src="arrow.svg"
        />
      </button>
    </div>
    <div
      className="btn "
    >
      <button
        className="btn-flat btn-md btn-primary"
        onClick={[Function]}
      >
        <img
          alt="arrow right"
          className="right"
          src="arrow.svg"
        />
      </button>
    </div>
  </div>
</div>
`;
